common/numerated/src/iterators.rs

```diff
@@ -0,0 +1,330 @@
+
+    fn try_from(range: (S, E)) -> Result<Self, Self::Error> {
+        match Interval::try_from(range) {
+            Ok(interval) => Ok(interval.into()),
+            Err(IncorrectOrEmptyRangeError::EmptyRange(_)) => Ok(Self(None)),
+            Err(IncorrectOrEmptyRangeError::IncorrectRange(err)) => Err(err),
+        }
+    }
+}
+
+impl<T: Numerated + UpperBounded, I: Into<T::Bound>> TryFrom<Range<I>> for IntervalIterator<T> {
+    type Error = IncorrectRangeError;
+
+    fn try_from(range: Range<I>) -> Result<Self, Self::Error> {
+        Self::try_from((range.start, range.end))
+    }
+}
+
+impl<T: Numerated> TryFrom<RangeInclusive<T>> for IntervalIterator<T> {
+    type Error = IncorrectRangeError;
+
+    fn try_from(range: RangeInclusive<T>) -> Result<Self, Self::Error> {
+        Interval::try_from(range).map(Into::into)
+    }
+}
+
+impl<T: Numerated> Iterator for IntervalIterator<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.map(|interval| {
+            let start = interval.start();
+            self.0 = interval.inc_start();
+            start
+        })
+    }
+}
+
+impl<T: Numerated + UpperBounded> IntervalIterator<T> {
+    /// Returns interval `start..start + len` if it's possible.
+    /// - if `len == None`, then it is supposed, that `len == T::Distance::max_value() + 1`.
+    /// - if `start + len - 1` is out of `T`, then returns [OutOfBoundsError].
+    /// - if `len` is zero, then returns empty interval.
+    pub fn new_with_len<S: Into<T::Bound>, L: Into<Option<T::Distance>>>(
+        start: S,
+        len: L,
+    ) -> Result<Self, OutOfBoundsError> {
+        match Interval::new_with_len(start, len) {
+            Ok(interval) => Ok(interval.into()),
+            Err(NewWithLenError::ZeroLen) => Ok(Self(None)),
+            Err(NewWithLenError::OutOfBounds(err)) => Err(err),
+        }
+    }
+}
+
+impl<T: Display> Display for IntervalIterator<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if let Some(interval) = &self.0 {
+            write!(f, "{}", interval)
+        } else {
+            write!(f, "∅")
+        }
+    }
+}
+
+/// Helper struct to iterate over intervals from `tree1`, which are not in `tree2`.
+///
+/// See also [`IntervalsTree::difference`].
+pub struct DifferenceIterator<T: Numerated, I: Iterator<Item = Interval<T>>> {
+    pub(crate) iter1: I,
+    pub(crate) iter2: I,
+    pub(crate) interval1: Option<Interval<T>>,
+    pub(crate) interval2: Option<Interval<T>>,
+}
+
+impl<T: Numerated, I: Iterator<Item = Interval<T>>> Iterator for DifferenceIterator<T, I> {
+    type Item = Interval<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let interval1 = if let Some(interval1) = self.interval1 {
+                interval1
+            } else {
+                let interval1 = self.iter1.next()?;
+                self.interval1 = Some(interval1);
+                interval1
+            };
```

Comment:
consider similar changes below
```suggestion
            self.interval1 = self.interval1.or_else(|| self.iter1.next());
            let interval1 = self.interval1?;
```

Reply:
done